File: Lab2/PartB/Shape.cpp
#include "Shape.h"
#include <iostream>
#include <string.h>
#include <math.h>

using namespace std;

Shape::Shape(int x, int y, string shapeName)
{
    this->origin = Point(x, y);
    this->shapeName = new char[shapeName.length() + 1];
    strcpy(this->shapeName, shapeName.c_str());
}

Shape::Shape(const Shape &source)
{
    origin = source.origin;
    shapeName = new char[strlen(source.shapeName) + 1];
    strcpy(shapeName, source.shapeName);
}

Shape &Shape::operator=(const Shape &rhs)
{
    if (this != &rhs)
    {
        origin = rhs.origin;
        delete[] shapeName;
        shapeName = new char[strlen(rhs.shapeName) + 1];
        strcpy(shapeName, rhs.shapeName);
    }
    return *this;
}

Shape::~Shape()
{
    delete[] shapeName;
}

const Point &Shape::getOrigin() const
{
    return origin;
}

const string Shape::getShapeName() const
{
    return shapeName;
}

void Shape::display() const
{
    cout << "Shape Name: " << shapeName << endl;
    origin.display();
}

double Shape::distance(Shape &other) const
{
    return origin.distance(other.origin);
}

double Shape::distance(Shape &theShape, Shape &other)
{
    return theShape.origin.distance(other.origin);
}

void Shape::move(double dx, double dy)
{
    origin.setx(origin.x() + dx);
    origin.sety(origin.y() + dy);
}

File: Lab4/ExD/EXD/shapes/Shape.java
//  Shape.java
package shapes;

public abstract class Shape {
	protected String name;
	protected String id;
	static int counter =0;
	Point origin;
	public abstract double area();
	public abstract double perimeter();
  public abstract double volume();
	
	
	public Shape(double x_origin, double y_origin, String name){
		
		origin = new Point(x_origin,y_origin);
		this.name = new String(name);
		counter++;
		id = "Abstract Parent";
	}
	
	public Object clone() throws CloneNotSupportedException {
		Shape obj = (Shape) super.clone();
		obj.origin = (Point) origin.clone();
		return obj;
	}

	public void setName(String name)
	{
		this.name = name;
	}
	
	public String getOid()
	{
		return id;
	}
	public Point  getOrigin()   
	{
		return origin;
	}
	
	public String  getName()   
	{
		return name;
	}  
	
	double  distance(   Shape  other)
	{
		return origin.distance(other.origin);
	}
	
	double  distance(   Shape  the_shape,    Shape  other)
	{
		return Point.distance(the_shape.origin, other.origin);
	}
	
	
	void  move(double dx, double dy)
	{
		origin.setx(origin.getx()+dx);
		origin.sety(origin.gety()+dy);
	}
	
	public String toString()
	{
		String s = "\nShape name: " + name + "\nShape Id: " + id + "\nOrigin: " + origin;
		return s;
	}
	
	public static int count()
	{
		return counter;
	}
	
}

